icu4c/source/test/intltest/localebuildertest.cpp

```diff
@@ -0,0 +1,1607 @@
+            } else {
+                // Unknow test method
+                errln("Unknown test case method: There is an error in the test case data.");
+                break;
+            }
+            if (U_FAILURE(status)) {
+                if (strcmp("X", TESTCASE[tidx][i]) == 0) {
+                    // This failure is expected
+                    break;
+                } else {
+                    errln("FAIL: U_ILLEGAL_ARGUMENT_ERROR at offset %d %s %s", i,
+                          " in test case: ", actions.c_str());
+                    break;
+                }
+            }
+            if (strcmp("T", method) == 0) {
+                break;
+            }
+        }  // while(true)
+    }  // for TESTCASE
+}
+
+void LocaleBuilderTest::TestLocaleBuilderBasic() {
+    LocaleBuilder bld;
+    bld.setLanguage("zh");
+    Verify(bld, "zh", "setLanguage('zh') should success but got Error: %s\n");
+
+    bld.setScript("Hant");
+    Verify(bld, "zh-Hant", "setScript('Hant') should success but got Error: %s\n");
+
+    bld.setRegion("SG");
+    Verify(bld, "zh-Hant-SG", "setRegion('SG') should success but got Error: %s\n");
+
+    bld.setRegion("HK");
+    bld.setScript("Hans");
+    Verify(bld, "zh-Hans-HK",
+           "setRegion('HK') and setScript('Hans') should success but got Error: %s\n");
+
+    bld.setVariant("revised");
+    Verify(bld, "zh-Hans-HK-revised",
+           "setVariant('revised') should success but got Error: %s\n");
+
+    bld.setUnicodeLocaleKeyword("nu", "thai");
+    Verify(bld, "zh-Hans-HK-revised-u-nu-thai",
+           "setUnicodeLocaleKeyword('nu', 'thai'') should success but got Error: %s\n");
+
+    bld.setUnicodeLocaleKeyword("co", "pinyin");
+    Verify(bld, "zh-Hans-HK-revised-u-co-pinyin-nu-thai",
+           "setUnicodeLocaleKeyword('co', 'pinyin'') should success but got Error: %s\n");
+
+    bld.setUnicodeLocaleKeyword("nu", "latn");
+    Verify(bld, "zh-Hans-HK-revised-u-co-pinyin-nu-latn",
+           "setUnicodeLocaleKeyword('nu', 'latn'') should success but got Error: %s\n");
+
+    bld.setUnicodeLocaleKeyword("nu", "");
+    Verify(bld, "zh-Hans-HK-revised-u-co-pinyin",
+           "setUnicodeLocaleKeyword('nu', ''') should success but got Error: %s\n");
+
+    bld.setUnicodeLocaleKeyword("co", "");
+    Verify(bld, "zh-Hans-HK-revised",
+           "setUnicodeLocaleKeyword('nu', '') should success but got Error: %s\n");
+
+    bld.setScript("");
+    Verify(bld, "zh-HK-revised",
+           "setScript('') should success but got Error: %s\n");
+
+    bld.setVariant("");
+    Verify(bld, "zh-HK",
+           "setVariant('') should success but got Error: %s\n");
+
+    bld.setRegion("");
+    Verify(bld, "zh",
+           "setRegion('') should success but got Error: %s\n");
+}
+
+void LocaleBuilderTest::TestSetLanguageWellFormed() {
+    // http://www.unicode.org/reports/tr35/tr35.html#unicode_language_subtag
+    // unicode_language_subtag = alpha{2,3} | alpha{5,8};
```

Comment:
except we decided to also support alpha{4} in ICU, right?

Please also test that.